### ISSUE TITLE
code upload?

### ISSUE THREAD

**Issue by antonyalexos** [StatMixedML/XGBoostLSS#7]
Hello, I have read the paper and I have to say that it is some really great work. Is it possible to upload the code as other have asked for it? You have 87 stars and 14 forks without uploading the code. You could help a lot of people..

**Comment by StatMixedML** [comment]
Thanks for your comment. 

Currently, I am still working on getting the package ready. Once I have the problem of simultaneous parameter estimation across distributional parameters solved, I will release the R-package.

**Comment by antonyalexos** [comment]
Isn't it a bit odd that you have completed the paper but not the code? Apart from that if an issue is not closed then you should close it. I mean uploading the code is still open for now.

**Comment by StatMixedML** [comment]
> Isn't it a bit odd that you have completed the paper but not the code

The code on which the paper is based upon is complete in the sense that it replicates the results of the paper. However, as I state in the readme-file "_... the public release of the package is somewhat delayed_." I am planning to release the code once it satisfies sound coding and software development standards + easy to use. Also, I need to work on the problem of simultaneous parameter estimation across distributional parameters. 

> Apart from that if an issue is not closed then you should close it.

Right, I`ve re-opend the issue.

**Comment by hiyamgh** [comment]
Hello @StatMixedML Thank you for this awesome repository. Could you please notify us here in this thread when the code is released ?

**Comment by t0504b** [comment]
Can you please tell what library to use in R, to run XGBoostLSS. Thanks!

**Comment by wrkhard** [comment]
Hello @StatMixedML I saw that you have closed your issue with multi-param optimization and mentioned that a commit would be coming soon. Do you know when we may be able to access xgboostlss from CRAN? Very much looking forward to using this method in an upcoming paper!

**Comment by StatMixedML** [comment]
@wrkhard Thanks for your interest.

I managed to get around the multi-parameter problem in Python, but still working on the release. So most likely, there will be no CRAN package since I haven't figured it out how to do it in R.

Maybe I can help you with applying XGBoostLSS to your specific needs if that is an option? Happy to, let me know.

**Comment by StatMixedML** [comment]
Closing this with the release of version 0.1.0

**Comment by kmedved** [comment]
Congratulations on the release of 0.1 @StatMixedML - eagerly testing.

**Comment by StatMixedML** [comment]
@kmedved 

Let me know in case you have a request on implementing a new distribution. Would be great to see it being used!

**Comment by kmedved** [comment]
Well, I'd love to see any kind of skewed continuous distribution (e.g., a Skewed-T distribution). That's currently an area where existing uncertainty-supporting GBDT models (ngboost, catboost, xgboost-distribution) are lacking. I've tried to add this to Ngboost before, but unfortunately the math is over my head.

For context, I am the creator of the DARKO basketball projection system, and I frequently need skewed distributions to model potential career outcomes. Any ability to capture skewed distributions would be much appreciated.

Thanks @StatMixedML

**Comment by bryorsnef** [comment]
> @wrkhard Thanks for your interest.
> 
> I managed to get around the multi-parameter problem in Python, but still working on the release. So most likely, there will be no CRAN package since I haven't figured it out how to do it in R.
> 
> Maybe I can help you with applying XGBoostLSS to your specific needs if that is an option? Happy to, let me know.

Have you thought of using reticulate to just link to the python package? 

Also, I would love to see beta and gamma distributions. I was interested in using XGBoostLSS in multiple imputation by chained equations, and having a good distribution for imputing positively only and bounded variables would be great.

**Comment by StatMixedML** [comment]
@kmedved Great honor to see XGBoostLSS being used as a potential model at DARKO!

Since I don't know the characteristics of your data, do you have any specific skewed Student-t in mind, since there are numerous ways of implementing it. Also, there are a number of other skews distributions available that might fit your data well. 

Would you need a continuous distribution defined on -infty < y < infty, or would a skewed distribution on y>0 such as the Box-Cox Cole and and Green distribution also work? 

To facilitate the choice of an appropriate distribution, I am planning to add an unconditional density estimation of several candidate distributions, that allows the user to compare their likelihoods. What that be of interest to you?

I am currently on vacation returning Thursday in a week. So I would try to add a suitable distribution by the weekend end of November.

**Comment by StatMixedML** [comment]
@bryorsnef Using the reticulate package is a great idea! Thanks for the suggestion.

However, before I release an R version, I plan to extend the functionality of the package first, such as adding more distributions, such as your suggestions (beta, gamma). But I strongly encourage you to open a pull request, adding the reticulate extension!

I try to be as responsive as possible to any user requests. Since I am currently on vacation though, adding more distributions might take a little. Hoping for your understanding.

**Comment by StatMixedML** [comment]
@kmedved @wrkhard  I have added the Box-Cox T distribution that is defined for y>0 and also allows for skewness and kurtosis.

**Comment by kmedved** [comment]
Thank you @StatMixedML - that's very exciting (apologies for the delay responding). My target data is mostly positive, or can be transformed easily to being positive, so a Box-Cox distribution may work well. I had previously been experimenting with trying to implement a Skew-T distribution via NGBoost, but ran into issues making it work, but this may work.

I will do some testing this week and report back. 

> To facilitate the choice of an appropriate distribution, I am planning to add an unconditional density estimation of several candidate distributions, that allows the user to compare likelihoods. What that be of interest to you?

This would be wonderful, yes. This would be hugely helpful functionality, especially for those of us like myself still learning about different distribution properties.

**Comment by StatMixedML** [comment]
@kmedved Since XGBoostlSS heavily relies on gradients and hessians, you may want to scale your target data y before training to stabilize the gradient/hessian estimation. It can be somehting very simple, say ```y/10``` or ``y/100```. This allows you to easily re-normalize the predictions afterwards and to keep the gradient/hessian bounded. Let me know if this helps.

**Comment by StatMixedML** [comment]
@kmedved I have added a stabilization function for the estimation of Gradients and Hessians and also updated all distributions accordingly.
